client/test/e2e.test.js

```diff
@@ -0,0 +1,131 @@
+import test from 'tape';
+import puppeteer from 'puppeteer';
+
+const width = 1920;
+const height = 1080;
+
+// Opens the chromium window
+function browser(width, height) {
+  return puppeteer.launch({
+    headless: true,
+    args: [`--window-size=${width},${height}`],
+    // slowMo: 20 // For testing
+  });
+}
+
+function sleep(ms) {
+  return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+// Clicks the add workspace button, types a name, and clicks the button
+async function create_workspace(p, name) {
+  await p.waitForSelector('#add-workspace');
+  await p.click('#add-workspace');
+  await p.waitForSelector('#workspace-name');
+  await p.focus('#workspace-name');
+  await p.keyboard.type(name);
+  await p.click('#create-workspace');
+}
+
+async function workspace_exists(p, name) {
+  let exists = false;
+
+  await p.waitForSelector('.v-list-item__title');
+  let workspaces = await p.evaluate(() => {
+    let titles = [];
+    let doc_nodes = document.querySelectorAll('.v-list-item__title');
+    for (let node of doc_nodes) {
+      titles.push(node.innerText);
+    }
+    return titles;
+  });
+
+  exists = workspaces.includes(name);
+
+  return exists;
+}
+
+async function element_exists(empty = false, element_type, p, name) {
+  let exists, list_is_empty, tables;
+
+  await p.waitForSelector('[data-title="Tables"] .ws-detail-empty-list ');
+  tables = await p.evaluate(element_type => {
+    let titles = [];
+    let doc_nodes = document.querySelectorAll('[data-title="' + element_type + '"] .ws-detail-empty-list ');
+    for (let node of doc_nodes) {
+      titles.push(node.innerText);
+    }
+    return titles;
+  }, element_type);
+
+  if (empty) {
+    list_is_empty = tables.includes('info There\'s nothing here yet...');
+    return list_is_empty;
+  } else {
+    exists = tables.includes(name);
+    return exists;
+  }
```

Comment:
It's a bit of an antipattern to pack two different tests into one function, dispatching between them based on a boolean argument. A better approach might be to abstract everything before this if-block into a helper function, then split this if-block into two functions, each of which calls the helper function, then includes one branch or the other. That will make this code block easier to read, and the usage of these functions much clearer at their point of use.

Reply:
I agree and I knew it as I wrote it, I was just having a mental block. I've split it out into 3 functions in b3f9d38